### PR TITLE
Feature: Improvements to `automaticallyAdjustKeyboardInsets`

### PR DESCRIPTION
**Pull request by isidoro98**
## Summary
This PR builds on top of #31402, which introduced the `automaticallyAdjustsScrollIndicatorInsets` functionality. It aims to fix one of RN's longstanding pain point regarding the keyboard.

The changes provide a better way of handling the `ScrollView` offset when a keyboard opens. Currently, when a keyboard opens we apply an **offset** to the `Scrollview` that matches the size of the keyboard. This approach is great if we are using an `InputAccessoryView` but if we have multiple `TextInputs` in a `ScrollView`; offsetting the content by the size of the keyboard doesn't yield the best user experience. 

## Changelog

[General] [Changed] - Changes offset behaviour when keyboard opens in `ScrollView` with `automaticallyAdjustsScrollIndicatorInsets`

## Test Plan

The videos below compare the current and proposed behaviors for the following code:

```jsx
<ScrollView
  automaticallyAdjustKeyboardInsets
  keyboardDismissMode="interactive">
  {[...Array(10).keys()].map(item => (
    <CustomTextInput placeholder={item.toString()} key={item} />
  ))}
</ScrollView>
```

<table>
<tr>
<th>Current behavior</th>
<th>Proposal</th>
</tr>
<tr>
<td><video src="https://user-images.githubusercontent.com/25139053/200194972-1ac5f1cd-2d61-4118-ad77-95c04d30c98d.mov" /></td>
<td><video src="https://user-images.githubusercontent.com/25139053/200194990-53f28296-be11-4a47-be70-cec917d7deb1.mov" /></td>
</tr>
</table>

As can be seen in the video, the **current behavior** applies an offset to the `ScrollView` content regardless of where the `TextInput` sits on the screen. 

The proposal checks if the `TextInput` will be covered by the keyboard, and only then applies an offset. The offset applied is not the full size of the keyboard but instead only the required amount so that the `TextInput` is a **specific** distance above the top of the keyboard (customizable using the new `bottomKeyboardOffset` prop). This achieves a less "jumpy" experience for the user.

The proposal doesn't change the behavior of the `ScrollView` offset when an `InputAccessory` view is used, since it checks if the `TextField` that triggered the keyboard is a **descendant** of the `ScrollView` or not.
<br></br>
## Why not use other existing solutions?

RN ecosystem offers other alternatives for dealing with a keyboard inside a ScrollView, such as a `KeyboardAvoidingView` or using third party libraries like `react-native-keyboard-aware-scroll-view`. But as shown in the recordings below, these solutions don't provide the smoothness or behavior that can be achieved with `automaticallyAdjustsScrollIndicatorInsets`.

<table>
<tr>
<th>KeyboardAvoidingView</th>
<th>rn-keyboard-aware-scroll-view</th>
</tr>
<tr>
<td><video src="https://user-images.githubusercontent.com/25139053/200195145-de742f0a-6913-4099-83c4-7693448a8933.mov" /></td>
<td><video src="https://user-images.githubusercontent.com/25139053/200195151-80745533-16b5-4aa0-b6cd-d01041dbd001.mov" /></td>
</tr>
</table>

As shown in the videos, the `TextInput` is hidden by the keyboard for a split second before becoming visible.

Code for the videos above:

```jsx
// KeyboardAvoidingView
<KeyboardAvoidingView
  style={{flex: 1, flexDirection: 'column', justifyContent: 'center'}}
  behavior="padding"
  enabled>
  <ScrollView>
    {[...Array(10).keys()].map(item => (
      <CustomTextInput placeholder={item.toString()} key={item} />
    ))}
  </ScrollView>
</KeyboardAvoidingView>
```

```jsx
// rn-keyboard-aware-scroll-view
<KeyboardAwareScrollView>
  {[...Array(10).keys()].map(item => (
    <CustomTextInput placeholder={item.toString()} key={item} />
  ))}
</KeyboardAwareScrollView>
```

### CODE REVIEW

File: React/Views/ScrollView/RCTScrollView.m [hunk 1]
```diff
@@ -324,11 +325,28 @@ - (void)_keyboardWillChangeFrame:(NSNotification *)notification
...
-  CGFloat contentDiff = endFrame.origin.y - beginFrame.origin.y;
-  if (self.inverted) {
-    newContentOffset.y += contentDiff;
+  UIResponder *firstResponder = [UIResponder currentFirstResponder];
+  if ([firstResponder isKindOfClass: [UITextField class]] && [(UITextField *) firstResponder isDescendantOfView:_scrollView]) {
+    UITextField *textField = [UIResponder currentFirstResponder];
+    CGRect textFieldFrame = [textField.superview convertRect:textField.frame toView:nil];
+    // [32] is a placeholder. It's the distance between the bottom of the textfield and the top of the keyboard
+    // What value should be used? Should it be customizable with a prop?
+    CGFloat textFieldBottom = textFieldFrame.origin.y + textFieldFrame.size.height + 32;
```
Comment (Pranav-yadav) [inline React/Views/ScrollView/RCTScrollView.m:334]: > For android (Java/Kotlin) we can use; [`WindowInsets`](https://developer.android.com/develop/ui/views/layout/sw-keyboard#java) api to get the height of the soft. keyboard

UIKit seems to have api's like [view.keyboardLayoutGuide](https://developer.apple.com/documentation/uikit/keyboards_and_input/adjusting_your_layout_with_keyboard_layout_guide)
and,
[UIKeyboardLayoutGuide](https://developer.apple.com/documentation/uikit/uikeyboardlayoutguide)

PS: Providing prop to adjust it manually seems nice customization but, using native api's to calculate the height would be more appropriate in this case.

**Comment by Pranav-yadav** [comment]
> For android (Java/Kotlin) we can use; [`WindowInsets`](https://developer.android.com/develop/ui/views/layout/sw-keyboard#java) api to get the height of the soft. keyboard

UIKit seems to have api's like [view.keyboardLayoutGuide](https://developer.apple.com/documentation/uikit/keyboards_and_input/adjusting_your_layout_with_keyboard_layout_guide)
and,
[UIKeyboardLayoutGuide](https://developer.apple.com/documentation/uikit/uikeyboardlayoutguide)

PS: Providing prop to adjust it manually seems nice customization but, using native api's to calculate the height would be more appropriate in this case. + hardcoding the height would be impact the UX

**Comment by isidoro98** [comment]
> > For android (Java/Kotlin) we can use; [`WindowInsets`](https://developer.android.com/develop/ui/views/layout/sw-keyboard#java) api to get the height of the soft. keyboard
> 
> > UIKit seems to have api's like [view.keyboardLayoutGuide](https://developer.apple.com/documentation/uikit/keyboards_and_input/adjusting_your_layout_with_keyboard_layout_guide)
> > and,
> > [UIKeyboardLayoutGuide](https://developer.apple.com/documentation/uikit/uikeyboardlayoutguide)
> 
> > PS: Providing prop to adjust it manually seems nice customization but, using native api's to calculate the height would be more appropriate in this case.
> 
> * hardcoding the height will impact the UX

Hi, the hardcoded value represents the distance between the bottom of the `TextInput` and the top of the Keyboard. Not the size of the keyboard

Please see the screenshot below for reference. IMO the default value should be 0 and we should allow the user to customize this distance.

**Comment by Pranav-yadav** [comment]
> Hi, the hardcoded value represents the distance between the bottom of the TextInput and the top of the Keyboard. Not the size of the keyboard

Got it.. Thanks. 😄

Then adding a prop for it and keeping its default value resonable; makes sense..

**Comment by efstathiosntonas** [comment]
I've just patched it into 0.68.5, works as a charm. I had to hard code the distance between the input and the keyboard because I was using an absolute positioned element at the bottom of the screen.

**Comment by isidoro98** [comment]
> I've just patched it into 0.68.5, works as a charm. I had to hard code the distance between the input and the keyboard because I was using an absolute positioned element at the bottom of the screen so I believe a prop would be great to adjust it.

Hey, glad to hear it worked! Do you have any suggestions regarding what this prop should be called? Looking for ideas :)

**Comment by efstathiosntonas** [comment]
`bottomOffset`, `bottomKeyboardOffset' 🤔

**Comment by isidoro98** [comment]
> `bottomOffset`, `bottomKeyboardOffset` , `bottomOffsetPadding` 🤔

Add the prop, named it `bottomKeyboardOffset`, thank you for the ideas!

**Comment by aliraza-noon** [comment]
@isidoro98 why did you close this ?

**Comment by casperstr** [comment]
This is super! Why is this closed?

**Pull request closed without merging**